File: companies/Nvidia/130.surrounded-regions.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    void solve(vector<vector<char>>& board) {
        /*
        Approach: any 'O' connected to the border cannot be captured. BFS from
        all border 'O' cells and mark them safe as 'S'. Then flip remaining 'O'
        cells to 'X' and restore safe cells back to 'O'.

        C++ notes: queue<pair<int,int>> is the FIFO structure for BFS over grid
        coordinates.
        Complexity: O(m*n) time, O(m*n) worst-case queue space.
        */
        if (board.empty() || board[0].empty()) return;
        int rows = board.size(), cols = board[0].size();
        queue<pair<int,int>> q;
        auto mark = [&](int r, int c) {
            if (board[r][c] == 'O') {
                board[r][c] = 'S';
                q.push({r, c});
            }
        };
        for (int r = 0; r < rows; ++r) { mark(r, 0); mark(r, cols - 1); }
        for (int c = 0; c < cols; ++c) { mark(0, c); mark(rows - 1, c); }
        int dirs[5] = {1, 0, -1, 0, 1};
        while (!q.empty()) {
            auto [r, c] = q.front(); q.pop();
            for (int d = 0; d < 4; ++d) {
                int nr = r + dirs[d], nc = c + dirs[d + 1];
                if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && board[nr][nc] == 'O') {
                    board[nr][nc] = 'S';
                    q.push({nr, nc});
                }
            }
        }
        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {
                if (board[r][c] == 'O') board[r][c] = 'X';
                else if (board[r][c] == 'S') board[r][c] = 'O';
            }
        }
    }
};

File: companies/EPAM Systems/130.surrounded-regions.java
import java.util.*;

/**
 * Algorithm:
 * Any 'O' connected to the border cannot be captured. BFS from border 'O'
 * cells and mark them safe as 'S'. Then flip remaining 'O' to 'X' and restore
 * safe cells back to 'O'.
 *
 * Java data structures:
 * ArrayDeque<int[]> is the BFS queue of board coordinates.
 *
 * Complexity:
 * Time O(rows * cols), space O(rows * cols) worst-case queue.
 */
class Solution {
    public void solve(char[][] board) {
        if (board == null || board.length == 0 || board[0].length == 0) {
            return;
        }
        int rows = board.length;
        int cols = board[0].length;
        Queue<int[]> q = new ArrayDeque<>();
        for (int r = 0; r < rows; r++) {
            mark(board, q, r, 0);
            mark(board, q, r, cols - 1);
        }
        for (int c = 0; c < cols; c++) {
            mark(board, q, 0, c);
            mark(board, q, rows - 1, c);
        }
        int[][] dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            for (int[] dir : dirs) {
                int nr = cur[0] + dir[0];
                int nc = cur[1] + dir[1];
                if (0 <= nr && nr < rows && 0 <= nc && nc < cols && board[nr][nc] == 'O') {
                    board[nr][nc] = 'S';
                    q.offer(new int[] {nr, nc});
                }
            }
        }
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (board[r][c] == 'O') {
                    board[r][c] = 'X';
                } else if (board[r][c] == 'S') {
                    board[r][c] = 'O';
                }
            }
        }
    }

    private void mark(char[][] board, Queue<int[]> q, int r, int c) {
        if (board[r][c] == 'O') {
            board[r][c] = 'S';
            q.offer(new int[] {r, c});
        }
    }
}

